src/sorunlib/wiregrid.py

```diff
@@ -198,6 +198,112 @@ def _check_temperature_sensors():
+                           + "LSR2: {}\n".format('ON' if ls_status['LSR2'] == 1 else 'OFF')
+                           + "LSL2: {}\n".format('ON' if ls_status['LSL2'] == 1 else 'OFF')
+                           + "Aborting...")
+    return position
+
+
+def _stop_hwp_with_wiregrid():
+    """Stop the HWP after comfirming the wiregrid is inserted."""
+    # Check the wiregrid position
+    position = _check_wiregrid_position()
+    if position == 'outside':
+        raise RuntimeError("The wiregrid is not inserted. Aborting...")
+    elif position == 'inside':
+        # Stop the HWP
+        print("Starting to stop the HWP.")
+        run.hwp.stop(active=True)
+        print("The HWP stopped successfully.")
+    else:
+        raise RuntimeError("Unknown wiregrid position. Aborting...")
+
+
+def _spin_hwp_with_wiregrid(target_hwp_direction):
+    """Spin the HWP to the target direction at 2Hz after comfirming the wiregrid is inserted.
+
+    Args:
+        target_hwp_direction (str): Target HWP direction, 'forward' or 'backward'.
+    """
+    # Check the wiregrid position
+    position = _check_wiregrid_position()
+    if position == 'outside':
+        raise RuntimeError("The wiregrid is not inserted. Aborting...")
+    elif position == 'inside':
+        if target_hwp_direction == 'forward':
+            print("Starting to spin up the HWP in forward.")
+            run.hwp.set_freq(freq=2.0)
+            print("The HWP is spinning at 2Hz in forward successfully.")
+        elif target_hwp_direction == 'backward':
+            print("Starting to spin up the HWP in backward.")
+            run.hwp.set_freq(freq=-2.0)
+            print("The HWP is spinning at 2Hz in backward successfully.")
+        else:
+            raise RuntimeError("Unknown HWP target direction. Aborting...")
+    else:
+        raise RuntimeError("Unknown wiregrid position. Aborting...")
+
+
+def _reverse_hwp_with_wiregrid(initial_hwp_direction, streaming=False, stepwise_before=False, stepwise_after=False):
+    """Change the HWP direction after comfirming the wiregrid is inserted.
+
+    Args:
+        initial_hwp_direction (str): Initial HWP direction, 'forward' or 'backward'.
+        streaming (bool): Do SMuRF streaming during the HWP direction chaging or not . Default is False.
+        stepwise_before (bool): Do stepwise rotation before changing HWP direction or not. Default is False.
+        stepwise_after (bool): Do stepwise rotation after changing HWP direction or not. Default is False.
+    """
+    if initial_hwp_direction not in ['forward', 'backward']:
+        raise RuntimeError("Initial HWP direction should be either 'forward' or 'backward'. Aborting...")
+
+    current_hwp_direction = initial_hwp_direction
+    if current_hwp_direction == 'forward':
+        target_hwp_direction = 'backward'
+    elif current_hwp_direction == 'backward':
+        target_hwp_direction = 'forward'
```

Comment:
I want to make sure we get this convention correct. The HWP supervisor agent has an argument called `--forward-dir` that defines 'forward' to be clockwise or counter-clockwise, which seems to imply to me this isn't really a fixed direction. Is there some document about what our convention should be?

(I tried to avoid this question myself when writing `hwp.set_freq()`, but it seems I can't escape it.)

Reply:
I found the confluence page about the rotation direction definition: [CHWP main page -- Rotation direction definition](https://simonsobs.atlassian.net/wiki/spaces/PRO/pages/185434251/CHWP+main+page#Rotation-direction-definition). The cw/ccw seems to correspond to the PID value depending on each satp and the PID corresponds to 'forward' or 'reverse' not depending on the satp. We should pay attention which physical direction the 'forward' and 'reverse' mean - cw or ccw. This document uses 'reverse' instead of 'backward'.
Also, I found the 'forwards' and 'backwards' (not 'reverse') in the observation plans on [the confluence](https://simonsobs.atlassian.net/wiki/spaces/SOPS/pages/736591902/Operations+2024-11-02+-+2024-11-08+SATp1). I guess a lot of people may make the 'forwards' correspond to the plus frequency like 'run.hwp.set_freq(freq=+2)`